include/zephyr/drivers/charger.h

```diff
@@ -0,0 +1,196 @@
+	/** External supply not presented */
+	CHARGER_ONLINE_OFFLINE = 0,
+	/** External supply is present and of fixed output */
+	CHARGER_ONLINE_FIXED,
+	/** External supply is present and of programmable output*/
+	CHARGER_ONLINE_PROGRAMMABLE,
+};
+
+/**
+ * @brief Charging states
+ */
+enum charger_status {
+	/** Charging device state is unknown */
+	CHARGER_STATUS_UNKNOWN = 0,
+	/** Charging device is charging a battery */
+	CHARGER_STATUS_CHARGING,
+	/** Charging device is not able to charge a battery */
+	CHARGER_STATUS_DISCHARGING,
+	/** Charging device is not charging a battery */
+	CHARGER_STATUS_NOT_CHARGING,
+	/** The battery is full and the charging device will not attempt charging */
+	CHARGER_STATUS_FULL,
+};
+
+/**
+ * @typedef charger_property_t
+ * @brief A charger property's identifier
+ *
+ * See charger_props for a list of identifiers
+ */
+typedef uint16_t charger_property_t;
+
+/**
+ * @brief Representation of a charger property
+ *
+ * The property has a property type that identifies the property to get,
+ * an error associated with fetching the property, and the fetched value
+ * of the property.
+ */
+struct charger_get_property {
+	/** Battery charger property to get */
+	charger_property_t property_type;
+	/** Negative error status set by callee e.g. -ENOTSUP for an unsupported property */
+	int err;
+	/** Property field for getting */
+	union {
+		enum charger_online online;
+		bool present;
+		enum charger_status status;
+	} value;
+};
+
+/**
+ * @typedef charger_get_properties_t
+ * @brief Callback API for getting multiple charger properties.
+ *
+ * See charger_get_property() for argument description
+ */
+typedef int (*charger_get_properties_t)(const struct device *dev,
+					struct charger_get_property *props, size_t props_len);
+
+/**
+ * @typedef charger_get_property_t
+ * @brief Callback API for getting a charger property.
+ *
+ * See charger_get_property() for argument description
+ */
+typedef int (*charger_get_property_t)(const struct device *dev, struct charger_get_property *prop);
+
+/**
+ * @brief Charging device API
+ *
+ * Caching is entirely on the onus of the client
+ */
+__subsystem struct charger_driver_api {
+	charger_get_property_t get_property;
+};
+
+/**
+ * @brief Fetch a battery charger property
+ *
+ * @param dev Pointer to the battery charger device
+ * @param prop pointer to charger_get_property struct where the property struct
+ * field is set by the caller to determine what property is read from the
+ * charger device into the charger_get_property struct's value field.
+ *
+ * @retval 0 if successful
+ * @retval < 0 if getting property failed
+ */
+__syscall int charger_get_prop(const struct device *dev, struct charger_get_property *prop);
+
+static inline int z_impl_charger_get_prop(const struct device *dev,
+					  struct charger_get_property *prop)
+{
+	const struct charger_driver_api *api = (const struct charger_driver_api *)dev->api;
+
+	return api->get_property(dev, prop);
+}
```

Comment:
Why not just take a `charger_property_t` input, and a `union ...` (to be typedeffed) output. This makes single prop clients life easier.

Reply:
I agree. The proposed API is not very "Zephyr-like".

Reply:
I'm okay either way but I don't think it's necessary for `charger_get_prop` for retrieving a singular property to exist if we can get 1 or more properties with `charger_get_props()` below.

Not hard requirement from me, but just a suggestion.